### PR TITLE
[CIS-746] Audit ChatMessageComposerSendButton

### PR DESCRIPTION
**Pull request by DominikBucher12**
# What this PR do:
- Audits `ChatMessageComposerSendButton`
# How to test it
- Run the tests, check if the implementation can be improved in some way
# Where you can start
- `ChatMessageComposerSendButton`

NOTE: Currently, the Button can be 100% customized only by subclassing because of the `mode` logic on `updateContent`... This might be a bit tricky, currently looking for a  better solution...

As part of this task I will audit `_ComposerButton` aka old `SquareButton` (Used for attachments/images and commands on the left side of the composer)

### CODE REVIEW

File: Sources/StreamChatUI/ChatMessageList/MessageComposer/ChatMessageComposerSendButton.swift [hunk 1]
```diff
@@ -5,19 +5,25 @@
 import StreamChat
 import UIKit
 
+/// Button used for final sending or confirming editing messages in ComposerView.
 public typealias ChatMessageComposerSendButton = _ChatMessageComposerSendButton<NoExtraData>
 
+/// Button used for final sending or confirming editing messages in ComposerView.
 open class _ChatMessageComposerSendButton<ExtraData: ExtraDataTypes>: _ComposerButton<ExtraData> {
```
Comment (DominikBucher12) [inline Sources/StreamChatUI/ChatMessageList/MessageComposer/ChatMessageComposerSendButton.swift:12]: Maybe add some documentation that complete customization can be achieved by subclassing due to `mode` behavior of this component? Or maybe rethink the behavior, I am trying to come up with something...

File: Sources/StreamChatUI/CommonViews/ComposerButton.swift [hunk 2]
```diff
@@ -5,15 +5,11 @@
...
 import UIKit
 
-public typealias ChatSquareButton = _ChatSquareButton<NoExtraData>
+public typealias ComposerButton = _ComposerButton<NoExtraData>
 
-open class _ChatSquareButton<ExtraData: ExtraDataTypes>: _Button, UIConfigProvider {
-    // MARK: - Properties
-    
+open class _ComposerButton<ExtraData: ExtraDataTypes>: _Button, UIConfigProvider {
     public var defaultIntrinsicContentSize: CGSize?
```
Comment (VojtaStavik) [inline Sources/StreamChatUI/CommonViews/ComposerButton.swift:11]: The size should be set by the superview that uses this view. There's no need for `defaultIntrinsicContentSize`...

File: Sources/StreamChatUI/CommonViews/ComposerButton.swift [hunk 3]
```diff
@@ -5,15 +5,11 @@
...
-    
-    // MARK: - Overrides
-    
+
```
Comment (VojtaStavik) [inline Sources/StreamChatUI/CommonViews/ComposerButton.swift:12]: Can't comment on it directly, but this whole thing should be here:
```swift
    override public func defaultAppearance() {
        defaultIntrinsicContentSize = .init(width: 40, height: 40)
        imageView?.contentMode = .scaleAspectFit
    }
```

File: Sources/StreamChatUI/CommonViews/ComposerButton.swift [hunk 4]
```diff
@@ -5,15 +5,11 @@
 import StreamChat
 import UIKit
 
-public typealias ChatSquareButton = _ChatSquareButton<NoExtraData>
+public typealias ComposerButton = _ComposerButton<NoExtraData>
 
-open class _ChatSquareButton<ExtraData: ExtraDataTypes>: _Button, UIConfigProvider {
-    // MARK: - Properties
-    
+open class _ComposerButton<ExtraData: ExtraDataTypes>: _Button, UIConfigProvider {
```
Comment (VojtaStavik) [inline Sources/StreamChatUI/CommonViews/ComposerButton.swift:10]: Which means, that this whole type is not needed :)

File: Sources/StreamChatUI/UIConfig+MessageComposer.swift [hunk 5]
```diff
@@ -20,8 +20,8 @@ public extension _UIConfig {
             _ChatMessageComposerImageAttachmentsView<ExtraData>.self
         public var documentAttachmentsView: _ChatMessageComposerDocumentAttachmentsView<ExtraData>.Type =
             _ChatMessageComposerDocumentAttachmentsView<ExtraData>.self
-        public var sendButton: _ChatMessageComposerSendButton<ExtraData>.Type = _ChatMessageComposerSendButton<ExtraData>.self
-        public var composerButton: _ChatSquareButton<ExtraData>.Type = _ChatSquareButton<ExtraData>.self
+        public var sendButton: _ChatMessageComposerSendButton.Type = _ChatMessageComposerSendButton<ExtraData>.self
+        public var composerButton: UIButton.Type = _ComposerButton<ExtraData>.self
```
Comment (VojtaStavik) [inline Sources/StreamChatUI/UIConfig+MessageComposer.swift:25]: Where is this used?

File: Sources/StreamChatUI/ChatMessageList/MessageComposer/ChatMessageComposerSendButton.swift [hunk 6]
```diff
@@ -7,23 +7,17 @@ import UIKit
 
 public typealias ChatMessageComposerSendButton = _ChatMessageComposerSendButton<NoExtraData>
 
-open class _ChatMessageComposerSendButton<ExtraData: ExtraDataTypes>: _ChatSquareButton<ExtraData> {
-    // MARK: Underlying types
-    
+open class _ChatMessageComposerSendButton<ExtraData: ExtraDataTypes>: _ComposerButton<ExtraData> {
```
Comment (VojtaStavik) [inline Sources/StreamChatUI/ChatMessageList/MessageComposer/ChatMessageComposerSendButton.swift:12]: Can't we just call it `SendButton` 🤷‍♂️

File: Sources/StreamChatUI/ChatMessageList/MessageComposer/ChatMessageComposerSendButton.swift [hunk 7]
```diff
@@ -5,19 +5,25 @@
...
 import UIKit
 
+/// Button used for final sending or confirming editing messages in ComposerView.
 public typealias ChatMessageComposerSendButton = _ChatMessageComposerSendButton<NoExtraData>
 
+/// Button used for final sending or confirming editing messages in ComposerView.
 open class _ChatMessageComposerSendButton<ExtraData: ExtraDataTypes>: _ComposerButton<ExtraData> {
+    /// Mode for the button
+    /// - new: When user wants to send new message.
+    /// - edit: When user edited his message and confirm changes.
```
Comment (VojtaStavik) [inline Sources/StreamChatUI/ChatMessageList/MessageComposer/ChatMessageComposerSendButton.swift:15]: Why not put the comments directly to the cases?

File: Sources/StreamChatUI/ChatMessageList/MessageComposer/ChatMessageComposerSendButton.swift [hunk 8]
```diff
@@ -5,19 +5,25 @@
...
     public enum Mode {
-        case new, edit
+        case new
+        case edit
```
Comment (VojtaStavik) [inline Sources/StreamChatUI/ChatMessageList/MessageComposer/ChatMessageComposerSendButton.swift:18]: can we be more expressive here? Maybe `Mode` -> `MessageKind` -> `newMessage` and `editedMessage`

File: Sources/StreamChatUI/ChatMessageList/MessageComposer/ChatMessageComposerSendButton.swift [hunk 9]
```diff
@@ -57,7 +63,7 @@ open class _ChatMessageComposerSendButton<ExtraData: ExtraDataTypes>: _ComposerB
             setImageWithAnimation(disabledStateImage, for: .disabled)
         }
     }
-    
+
     private func setImageWithAnimation(_ image: UIImage?, for state: UIControl.State) {
```
Comment (VojtaStavik) [inline Sources/StreamChatUI/ChatMessageList/MessageComposer/ChatMessageComposerSendButton.swift:67]: This should be `open`

**Review by VojtaStavik — CHANGES_REQUESTED** [review]
Looks like a good start but IMO it deserves a bit more love and more radical approach.

**Pull request closed without merging**